### PR TITLE
developing back end run methods

### PR DESCRIPTION
**Pull request by cadeduckworth**
* separating functionality for use with _single_universe without iterating frames
* retaining functionality for use with _single_frame

### CODE REVIEW

File: mdpow/analysis/ensemble.py [hunk 1]
```diff
@@ -516,16 +516,17 @@ def run(self, start=None, stop=None, step=None):
...
+            except NotImplementedError:
+                for i, ts in enumerate(ProgressBar(self._trajectory[self.start:self.stop:self.step], verbose=True,
                                                postfix=f'running system {self._key}')):
                 self._frame_index = i
                 self._ts = ts
                 self.frames[i] = ts.frame
                 self.times[i] = ts.time
                 self._single_frame()
-            self._conclude_universe()
+                raise
```
Comment (orbeckst) [inline mdpow/analysis/ensemble.py:529]: why raise??

where is conclude?

File: mdpow/analysis/ensemble.py [hunk 2]
```diff
@@ -516,16 +516,17 @@ def run(self, start=None, stop=None, step=None):
         self._prepare_ensemble()
         for self._key in ProgressBar(self._ensemble.keys(), verbose=True):
             self._setup_system(self._key, start=start, stop=stop, step=step)
-            self._prepare_universe()
-            self._single_universe()
-            for i, ts in enumerate(ProgressBar(self._trajectory[self.start:self.stop:self.step], verbose=True,
+            try:
+                self._single_universe()
+            except NotImplementedError:
```
Comment (orbeckst) [inline mdpow/analysis/ensemble.py:521]: Need to document that NotImplementedError is a sign to do single frame analysis.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Please address comments and add tests + CHANGELOG.

File: mdpow/analysis/ensemble.py (hunk 1 above)
Comment (cadeduckworth) [reply mdpow/analysis/ensemble.py:529]: Based on your discussion with Alia it was decided that conclude universe was no longer necessary for the functionality before or after splitting the back end run methods. I will try to find the conversation and post a screenshot here shortly.

Comment (cadeduckworth) [reply mdpow/analysis/ensemble.py:529]: https://github.com/Becksteinlab/MDPOW/pull/199#issuecomment-1241307425

Comment (orbeckst) [reply mdpow/analysis/ensemble.py:536]: Ok, that’s goid then.

Comment (orbeckst) [reply mdpow/analysis/ensemble.py:536]: But why the raise?

Comment (cadeduckworth) [reply mdpow/analysis/ensemble.py]: Sorry, I removed raise again and fixed the expected indent error, so it should be okay now. I could not figure out how to get it to work properly after making the changes, but I think the error is in how my development environment is configured, or how _single_universe() is defined in the EnsembleAnalysisClass I was using to test. The original changes were made using an older version of Python, so I am going to start my development environment over from scratch, redo the usage testing with a simple analysis class, and see if any other changes are required.

Comment (orbeckst) [reply mdpow/analysis/ensemble.py]: Your version of Python or IDE should not affect how Python code is interpreted. As long you use Python ≥ 3.7 everything should be fine.

File: mdpow/analysis/ensemble.py (hunk 2 above)
Comment (cadeduckworth) [reply mdpow/analysis/ensemble.py:528]: Unfortunately I think this method will not work based on how the EnsembleAnalysis base class is written in ensemble.py or how inheritance is being passed to DihedralAnalysis. 

`    def _single_universe(self):
        """Calculations on a single Universe object.
            Run on each universe in the ensemble during when
            self.run in called.
        """
        pass  # pragma: no cover

    def _single_frame(self):
        """Calculate data from a single frame of trajectory
        Called on each frame for universes in the Ensemble.
        """
        pass  # pragma: no cover`

This allowed for both self._single_universe() and self._single_frame() to be called in the same block every  time .run() was called by an EnsembleAnalysis subclass (pre try/except), which means NotImplementedError is never being raised even when _single_universe() is only defined in the base class and not the subclass. 

For example, with the try/except method as it is here now, when running DihedralAnalysis, self._single_universe() is called, nothing happens because it is only defined with 'pass' in the base class, NotImplementedError is not raised, and the except block is not executed. Therefore, self._single_frame() is never called, and no data is analyzed or stored in the results DataFrame expected from DihedralAnalysis. I tested this on solvation shell analysis, which also uses _single_frame(), and got the same problem, an empty results DataFrame.

I am looking into alternative options that don't require modifying anything outside of the run method.

Comment (orbeckst) [reply mdpow/analysis/ensemble.py:528]: Modify the base class and `raise NotImplementedError` instead of `pass`. This will ensure that developers implement at least one of the two.

File: mdpow/analysis/ensemble.py [hunk 3]
```diff
@@ -469,14 +469,14 @@ def _single_universe(self):
...
-        pass  # pragma: no cover
+        raise NotImplementedError
 
     def _single_frame(self):
         """Calculate data from a single frame of trajectory
 
         Called on each frame for universes in the Ensemble.
         """
-        pass  # pragma: no cover
+        raise NotImplementedError
```
Comment (orbeckst) [inline mdpow/analysis/ensemble.py:479]: Add test that checks that this NotImplementedError is raised. If necessary by calling the method explicitly
```python
def test_single_frame_raises_NotImplementedError():
    ...
    ea = EnsembleAnalysis(...)
    with pytest.raises(NotImplementedError):
          ea._single_frame()
```

File: mdpow/analysis/ensemble.py [hunk 4]
```diff
@@ -505,27 +505,34 @@ def _conclude_ensemble(self):
...
         First iterates through keys of ensemble, then runs _setup_system
         which defines the system and trajectory. Then iterates over
         trajectory frames.
+        
+        NotImplementedError will detect whether _single_universe or _single_frame
+        should be implemented, based on which is defined in the EnsembleAnalysisClass.
+        Only one of the two aforementioned functions should be defined for the respective
+        analysis class. For verbose functionality, the analysis will currently show two
+        iteration bars, where only one of which will actually be iterated, while the other
+        will load to completion instantaneously, showing the system that is being worked on.
```
Comment (orbeckst) [inline mdpow/analysis/ensemble.py:520]: These implementation details are developer documentation and do not need to be in `run()` — users read these docs and will be confused. Put it into the part of the docs that talks about EnsembleAnalysis and how to write your own. Add a section to https://mdpow.readthedocs.io/en/latest/analysis/ensemble_analysis.html

File: mdpow/tests/test_dihedral.py [hunk 5]
```diff
@@ -23,9 +23,9 @@
 
 class TestDihedral(object):
     DG48910_mean = -172.9849512527183
-    DG491011_mean = -3.7300197060478695
-    DG48910_var = 1490.6576365537262
-    DG491011_var = 128.3805265432388
+    DG491011_mean = 177.74725233051953
+    DG48910_var = 0.20311120667628546
+    DG491011_var = 0.006976126708773456
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:28]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 6]
```diff
@@ -37,25 +37,25 @@ def teardown(self):
         self.tmpdir.dissolve()
 
     def test_dataframe(self):
-        dh1 = self.Ens.select_atoms('name C4 or name C17 or name S2 or name N3')
+        dh1 = self.Ens.select_atoms('name C4', 'name C17', 'name S2', 'name N3')
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:40]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 7]
```diff
@@ -37,25 +37,25 @@ def teardown(self):
...
         dh_run = DihedralAnalysis([dh1]).run(start=0, stop=4, step=1)
 
         results = dh_run.results
 
-        assert results['selection'][0] == 'S2-N3-C4-C17'
+        assert results['selection'][0] == 'C4-C17-S2-N3'
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:45]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 8]
```diff
@@ -37,25 +37,25 @@ def teardown(self):
...
         for s in results['solvent']:
             assert s == 'water'
         for i in results['interaction'][:12]:
             assert i == 'Coulomb'
 
     def test_selection_error(self):
-        dh1 = self.Ens.select_atoms('name C17 or name S2 or name N3')
+        dh1 = self.Ens.select_atoms('name C17', 'name S2', 'name N3')
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:52]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 9]
```diff
@@ -37,25 +37,25 @@ def teardown(self):
...
         with pytest.raises(SelectionError):
             dh_run = DihedralAnalysis([dh1]).run(start=0, stop=4, step=1)
 
     def test_results_recursive1(self):
-        dh1 = self.Ens.select_atoms('name C11 or name C10 or name C9 or name C4')
-        dh2 = self.Ens.select_atoms('name C11 or name C10 or name C9 or name C4')
+        dh1 = self.Ens.select_atoms('name C11', 'name C10', 'name C9', 'name C4')
+        dh2 = self.Ens.select_atoms('name C11', 'name C10', 'name C9', 'name C4')
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:58]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 10]
```diff
@@ -64,29 +64,34 @@ def test_results_recursive1(self):
             assert dh_run1.results['dihedral'][i] == dh_run2.results['dihedral'][i]
 
     def test_results_recursive2(self):
-        dh1 = self.Ens.select_atoms('name C11 or name C10 or name C9 or name C4')
-        dh2 = self.Ens.select_atoms('name C8 or name C4 or name C9 or name C10')
+        dh1 = self.Ens.select_atoms('name C11', 'name C10', 'name C9', 'name C4')
+        dh2 = self.Ens.select_atoms('name C8', 'name C4', 'name C9', 'name C10')
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:68]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 11]
```diff
@@ -64,29 +64,34 @@ def test_results_recursive1(self):
...
 
         dh_run = DihedralAnalysis([dh1, dh2]).run(start=0, stop=4, step=1)
 
-        dh1_result = dh_run.results.loc[dh_run.results['selection'] == 'C4-C9-C10-C11']['dihedral']
-        dh2_result = dh_run.results.loc[dh_run.results['selection'] == 'C4-C8-C9-C10']['dihedral']
+        dh1_result = dh_run.results.loc[dh_run.results['selection'] == 'C11-C10-C9-C4']['dihedral']
+        dh2_result = dh_run.results.loc[dh_run.results['selection'] == 'C8-C4-C9-C10']['dihedral']
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:73]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 12]
```diff
@@ -64,29 +64,34 @@ def test_results_recursive1(self):
...
         dh2_var = circvar(dh2_result, high=180, low=-180)
 
-        assert_almost_equal(self.DG48910_mean, dh1_mean, 6)
-        assert_almost_equal(self.DG48910_var, dh1_var, 6)
-        assert_almost_equal(self.DG491011_mean, dh2_mean, 6)
-        assert_almost_equal(self.DG491011_var, dh2_var, 6)
+        assert_almost_equal(dh1_mean, self.DG48910_mean, 6)
+        assert_almost_equal(dh1_var, self.DG48910_var, 6)
+        assert_almost_equal(dh2_mean, self.DG491011_mean, 6)
+        assert_almost_equal(dh2_var, self.DG491011_var, 6)
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:83]: revert changes, will be fixed in PR #218

File: mdpow/tests/test_dihedral.py [hunk 13]
```diff
@@ -64,29 +64,34 @@ def test_results_recursive1(self):
...
 
     def test_ValueError_different_ensemble(self):
         other = Ensemble(dirname=self.tmpdir.name, solvents=['water'])
-        dh1 = self.Ens.select_atoms('name C11 or name C10 or name C9 or name C4')
-        dh2 = other.select_atoms('name C8 or name C4 or name C9 or name C10')
+        dh1 = self.Ens.select_atoms('name C11', 'name C10', 'name C9', 'name C4')
+        dh2 = other.select_atoms('name C8', 'name C4', 'name C9', 'name C10')
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:88]: revert changes, will be fixed in PR #218

**Review by orbeckst — CHANGES_REQUESTED** [review]
There are some edits that are now being  fixed in PR #218 so don't make the changes here and wait for #218 to be merged.

See other inline comments but overall looking good!

File: mdpow/analysis/ensemble.py (hunk 3 above)
Comment (cadeduckworth) [reply mdpow/analysis/ensemble.py:479]: https://github.com/Becksteinlab/MDPOW/blob/ensemble_run_update/mdpow/tests/test_dihedral.py
- file does not exist in pull request because it was deleted and moved to another pull request, so I just linked the source from the branch
- function is at end of module
- this is the explicit call of _single_universe() to test NotImplementedError is raised when not defined, but specifically moved for us in DihedralAnalysis as instructed

I am going to add explicit tests for raising NotImplementedError for _single_frame() and _single_universe() similar to pattern in your comment above, placed in test_ensemble.py, because test_run.py is for the partition coefficient calculation and test_ensemble.py deals with EnsembleAnalysis class where this run method that calls single_frame/uni is defined
- will be in this pull request

File: mdpow/tests/test_dihedral.py (hunk 5 above)
Comment (cadeduckworth) [reply mdpow/tests/test_dihedral.py:28]: new values may still required, circvar should always be between 0 and 1 per the scipy docs

File: mdpow/tests/test_dihedral.py (hunk 6 above)
Comment (cadeduckworth) [reply mdpow/tests/test_dihedral.py:40]: with 'or' vs. ', ' the selection string is reordered alphanumerically, dihedral may be calculated incorrectly if atom group is reordered

File: mdpow/tests/test_dihedral.py (hunk 7 above)
Comment (cadeduckworth) [reply mdpow/tests/test_dihedral.py:45]: selection order consistent with original selection string

Comment (orbeckst) [reply mdpow/tests/test_dihedral.py]: My point was that each PR should address _separate_ issues and there should not be any duplicated code changes. PRs should be "orthogonal" to each other and not contain each other's work. In this case it does not seem to matter because I was able to merge PR #218 , which properly fixes the problem, into _this_ PR without problem but this is not always straightforward. And conceptually it's messy: It's best when one PR fixes one problem.

File: mdpow/tests/test_dihedral.py (hunk 6 above)
Comment (orbeckst) [reply mdpow/tests/test_dihedral.py]: My point was that there should not be duplicated code across PRs. **One PR solves ONE problem.** is a good mantra. I would have expected you to just `git revert <SHA1 commit ID>` the changes so that this PR would have retained the incorrect ordering and reference values. As soon as PR #218 was merged, we would have used the ''update this PR with current develop'' and merge develop with the good changes to test_recursive2() into this PR. I did this anyway and and the merge produced clean code without conflicts (so when you look into the file changes for this PR it appears as if you never touched test_recursive2() in this PR), but that's not always guaranteed.

In short: avoid duplicating code across PRs. If you need something for a certain PR to work then that's a sign that all effort needs to go into making this other thing work.

Comment (cadeduckworth) [reply mdpow/tests/test_dihedral.py]: My apologies, I think I tried to merge the branches and did so locally, but did not realize I could not merge and close the other branch remotely without the review being addressed, which resulted in duplicated code changes across branches after pushing the changes. I will continue working to close the gap in my knowledge of proper use of git when dealing with multiple related branches.

File: mdpow/analysis/ensemble.py (hunk 4 above)
Comment (cadeduckworth) [reply mdpow/analysis/ensemble.py:520]: Do I need to create a Read the Docs account and link my GitHub account and repository to make these changes as indicated in the Sphinx documentation, or do I make the changes to the text files (i.e. ensemble_analysis.txt) and push the changes straight to GitHub?

Comment (orbeckst) [reply mdpow/analysis/ensemble.py:520]: You don't need an account. The CI is already using RTD.

You only need to edit the docs here, namely ensemble_analysis.txt.

Comment (orbeckst) [reply mdpow/analysis/ensemble.py:520]: The CI puts an updated view of the docs for the PR at https://mdpow--216.org.readthedocs.build/en/216/

File: CHANGES [hunk 14]
```diff
@@ -5,7 +5,26 @@ CHANGES for MDPOW
 Add summary of changes for each release. Use ISO 8061 dates. Reference
 GitHub issues numbers and PR numbers.
 
+2022-12-07    0.8.1
```
[the remark this answers is not in the document]
Comment (orbeckst) [reply CHANGES:8]: turn the date into 2022-??-?? — we don't know a release date yet

File: mdpow/analysis/ensemble.py (hunk 4 above)
Comment (cadeduckworth) [reply mdpow/analysis/ensemble.py]: Thank you! I think everything merged and pushed correctly, and all tests passed. I do need to modify the docs slightly after following the link above, but will wait until your review to make all necessary documentation changes in one commit.

File: CHANGES (hunk 14 above)
[the remark this answers is not in the document]
Comment (orbeckst) [reply CHANGES:8]: In fact: add everything to the 0.8.1 entry *below* and change that 0.8.1 to 0.9.0

File: doc/sphinx/source/analysis/ensemble_analysis.txt [hunk 15]
```diff
@@ -15,10 +15,17 @@ Universes and be analyzed as a group.
 
 :class:`~mdpow.analysis.ensemble.EnsembleAnalysis` is a class inspired by the
 :class:`AnalysisBase <MDAnalysis.analysis.base.AnalysisBase>` from MDAnalysis which
-iterates over the systems in the ensemble and the frames in the systems. It sets up both iterations between
-universes and universe frames allowing for analysis to be run on both whole systems and the frames of those
+iterates over the systems in the ensemble or the frames in the systems. It sets up iterations between
+universes or universe frames allowing for analysis to be run on either whole systems or the frames of those
 systems. This allows for users to easily run analyses on MDPOW simulations.
 
+NotImplementedError will detect whether _single_universe or _single_frame
```
[the remark this answers is not in the document]
Comment (orbeckst) [reply doc/sphinx/source/analysis/ensemble_analysis.txt:22]: To actually make the markup link to the doc, use something like
```reST
:meth:`~EnsembleAnalysis._single_universe`
```
The tilde will make it appear as only `_single_universe`.

File: mdpow/analysis/ensemble.py [hunk 16]
```diff
@@ -509,15 +517,8 @@ def run(self, start=None, stop=None, step=None):
         on each frame in the system.
 
         First iterates through keys of ensemble, then runs _setup_system
```
Comment (orbeckst) [inline mdpow/analysis/ensemble.py:519]: add markup
```reST
:meth:`_setup_system`
```

**Review by orbeckst — CHANGES_REQUESTED** [review]
Please see comments for guidance on doc writing.

Also need tests.

File: mdpow/tests/test_ensemble.py [hunk 17]
```diff
@@ -149,6 +149,21 @@ def _conclude_universe(self):
...
+            def test_single_frame(self):
+                with pytest.raises(NotImplementedError) as excinfo:
+                    TestAnalysis._single_frame(self)
+                assert excinfo.type == 'NotImplementedError'
+
+            def test_single_universe(self):
+                with pytest.raises(NotImplementedError) as excinfo:
+                    TestAnalysis._single_universe(self)
+                assert excinfo.type == 'NotImplementedError'
+
```
Comment (cadeduckworth) [inline mdpow/tests/test_ensemble.py:166]: #216 https://github.com/Becksteinlab/MDPOW/pull/216#discussion_r1042832663

@orbeckst Can you provide more detail for the linked comment? This test might not have shown up with the recent commit, but if it did, can you elaborate on how it is not correctly testing the base class?

Comment (orbeckst) [reply mdpow/tests/test_ensemble.py:166]: Codecov indicates that these lines were not tested. Codecov is not always right but it's something to look into carefully. Come up with a way to convince yourself (and me) that your tests really test the code that you think they do. Perhaps temporarily change the raise NotImplementError to ValueError and then your test should fail. Report back on what you did.

File: mdpow/analysis/ensemble.py (hunk 3 above)
Comment (orbeckst) [reply mdpow/analysis/ensemble.py:487]: Look into https://github.com/Becksteinlab/MDPOW/pull/216/files and you see the annotation by codecov that says that the line was not covered by tests.

Also check the codecov report in the checks and ultimately https://app.codecov.io/gh/Becksteinlab/MDPOW/pull/216

File: mdpow/tests/test_ensemble.py (hunk 17 above)
Comment (cadeduckworth) [reply mdpow/tests/test_ensemble.py:183]: I changed exception/error types and nothing changed. If the test and class definitions were nested correctly, pytest would have also picked up that 
`assert excinfo.type ==` should have been `assert excinfo.type **is**`, so basically nothing was happening. There is a way to match directly to string but I did it incorrectly and this way is simpler.

I split the single test into two separate, one for `_singe_frame` and one for `_single_universe`. Each has its own TestAnalysis class definition, where it shows that the run method not being tested is defined, so the undefined run method is explicitly tested for raising NotImplementedError, for each case. 

There is probably a way to consolidate the two tests back into one, but it is very clear what is being done this way. It still seems there is a simpler method.  I will review again in the morning and make changes requested for documentation so this can be merged if everything is correct otherwise!

**Comment by orbeckst** [comment]
Now coverage of the patch is 100%, that's good. Please ping me when you have addressed the remaining comments (docs, CHANGES, etc), @cadeduckworth . Thanks!

File: CHANGES [hunk 18]
```diff
@@ -5,13 +5,20 @@ CHANGES for MDPOW
...
 
+* for ensemble.EnsembleAnalysis._single_frame() 
+  changed 'pass' to 'raise NotImplementedError' (#216)
+* for ensemble.EnsembleAnalysis._single_universe() 
+  changed 'pass' to 'raise NotImplementedError' (#216)
+* for ensemble.EnsembleAnalysis.run() changed to detect 
+  and use _single_frame OR _single_universe (#216)
+* _prepare_universe and _conclude_universe removed from 
+  EnsembleAnalysis.run() method, no longer needed (per comments, #199)
+
```
Comment (orbeckst) [inline CHANGES:21]: remove empty line
```suggestion
```

File: mdpow/analysis/dihedral.py [hunk 19]
```diff
@@ -78,7 +78,7 @@ def check_dihedral_inputs(selections):
         for group in selections:
             for k in group.keys():
                 if len(group[k]) != 4:
-                    msg = ''''Dihedral calculations require AtomGroups with
+                    msg = '''Dihedral calculations require AtomGroups with
                               only 4 atoms, %s selected''' % len(group)
```
Comment (orbeckst) [inline mdpow/analysis/dihedral.py:82]: reformat this message so that it does not contain newline/space
```suggestion
                    msg = ("Dihedral calculations require AtomGroups with "
                           f"only 4 atoms, {len(group)} selected")
```
And we can use f-strings.

File: mdpow/analysis/ensemble.py [hunk 20]
```diff
@@ -464,19 +464,34 @@ def _setup_frames(self, trajectory):
...
+
+           Run on each :class:`MDAnalysis.Universe <MDAnalysis.core.groups.universe.Universe>` 
+           in the :class:`~mdpow.analysis.ensemble.Ensemble` 
+           during when :meth:`run` in called.
+            
+           :exc:`NotImplementedError` will detect whether 
+           :meth:`~EnsembleAnalysis._single_universe`
+           or :meth:`~EnsembleAnalysis._single_frame` 
+           should be implemented, based on which is defined 
+           in the :class:`~mdpow.analysis.ensemble.EnsembleAnalysis`.
```
Comment (orbeckst) [inline mdpow/analysis/ensemble.py:477]: Can you format it so that it fits into 79 character wide lines?

The first line can be broken over multiple lines, there should just be an empty line to separate the summary/header line from the synopsis.

File: mdpow/tests/test_dihedral.py [hunk 21]
```diff
@@ -93,3 +93,8 @@ def test_ValueError_different_ensemble(self):
                            match='Dihedral selections from different Ensembles, '):
             DihedralAnalysis([dh1, dh2])
 
+    def test_single_universe(self):
+        dh = self.Ens.select_atoms('name C4', 'name C17', 'name S2', 'name N3')
+        with pytest.raises(NotImplementedError):
+            DihedralAnalysis([dh])._single_universe()
+
```
Comment (orbeckst) [inline mdpow/tests/test_dihedral.py:100]: This test is not strictly necessary but we can leave it in, as a check that the DihedralAnalysis does not do something weird to `_single_universe()`.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Looking good, just minor comments.

**Review by orbeckst — APPROVED** [review]

**Comment by orbeckst** [comment]
Congratulations @cadeduckworth , PR is merged! 🚀
